oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/version/VersionableState.java

```diff
@@ -285,8 +287,32 @@ private void restoreFrozen(@NotNull NodeBuilder frozen,
                                @NotNull NodeBuilder dest,
                                @NotNull VersionSelector selector)
             throws RepositoryException, CommitFailedException {
-        // 15.7.2 Restoring Type and Identifier
-        restoreFrozenTypeAndUUID(frozen, dest);
+        //OAK-9459: if the NodeState is not empty, retrieve OPVs before restoring types to avoid constraint violations
+        boolean frozenTypeRestored = false;
+        if (!dest.hasProperty(JCR_PRIMARYTYPE)) {
+            // empty NodeState
+            // 15.7.2 Restoring Type and Identifier
+            restoreFrozenTypeAndUUID(frozen, dest);
+            frozenTypeRestored = true;
+        }
+        HashMap<PropertyState, Integer> opvs = new HashMap<>();
```

Comment:
```suggestion
        Map<PropertyState, Integer> opvs = new HashMap<>();
```

Reply:
It's completely irrelevant here, but as you wish.